### PR TITLE
[dataflow] TopK tidy and bugfix

### PR DESCRIPTION
**Pull request by frankmcsherry**
This PR adds a `region` around the dataflow fragment rendered for `TopK`, and also corrects an apparent error when a monotonic input has an `offset > 0` (we would retract records that precede the offset, mistaking them for records after `limit`). The first commit just adds a region, but is a bit noisy as some of the helper methods had opinions on the exact type of scope they would apply to.

@philip-stoev : I wonder if we should try and get more coverage on monotonic/append-only inputs?

### CODE REVIEW

**Comment by frankmcsherry** [comment]
Reviewing recommendation: toggle on "hide whitespace changes" to dramatically reduce the diff.

File: src/dataflow/src/render/top_k.rs [hunk 1]
```diff
@@ -40,38 +40,43 @@ where
...
+                    // This optimization generalizes to any TopK over a monotonic source, but we special
+                    // case only TopK with offset=0 and limit=1 (aka Top1) for now. This is because (1)
+                    // Top1 can merge in each incremental row in constant space and time while the
+                    // generalized solution needs something like a priority queue and (2) we expect Top1
+                    // will be a common pattern used to turn a Kafka source's "upsert" semantics into
+                    // differential's semantics. (2) is especially interesting because Kafka is
+                    // monotonic with an ENVELOPE of NONE, which is the default for ENVELOPE in
+                    // Materialize and commonly used by users.
+                    render_top1_monotonic(ok_input, group_key, order_key)
+                } else if *monotonic && *offset == 0 {
```
Comment (danhhz) [inline src/dataflow/src/render/top_k.rs:61]: Comment explaining what's going on with the `*offset == 0` plz

**Comment by danhhz** [comment]
I'm also inclined to add a regression test in this PR but it sounds like you're asking Philip to do it in a followup?

**Comment by frankmcsherry** [comment]
I'm happy to add a test, but I don't know that we have a framework for append-only inputs. It relies on a certain `ENVELOPE` setting for the input (`NONE`) and I think by default `sqllogictest` and `testdrive` don't present it. Or at least, I'm not familiar with the right idioms for it, and wanted to avoid making up something bespoke.

**Comment by danhhz** [comment]
Yeah, I hit this too in #6471. In that PR I used a VALUES, which is treated as monotonic, but I don't see how that would work here. Maybe just file a "regression test for 6885" issue then so it doesn't get dropped

**Comment by frankmcsherry** [comment]
Will do. I might go more general and ask for append-only testing, an instance of which would be a test for this.

**Comment by philip-stoev** [comment]
@frankmcsherry see existing test in https://github.com/MaterializeInc/materialize/blob/main/test/testdrive/top-1-monotonic.td for an example on NONE sources with the TopK optimization.

If you can figure out how to extend this test to hit the particular wrong result issue you are describing, that would be much apprecited.

**Comment by frankmcsherry** [comment]
I think it may be tricky to hit without a CDC source (or building in a 10s wait to the test). What happens is that the correct result is initially presented, but then 10s later results that precede `offset` are retracted from the input, which cause all records to move up, and some correct results to be removed because they new precede `offset`.

This could be tested with `AS OF` on select queries, but I think only CDC sources support this without actually waiting for time to pass. In the current code, there is a 10s latency for the "garbage collection" that triggers the issue.

File: src/dataflow/src/render/top_k.rs (hunk 1 above)
Comment (frankmcsherry) [reply src/dataflow/src/render/top_k.rs:61]: Added, and a TODO about how it could be relaxed.

**Review by danhhz — APPROVED** [review]
What's the culture here? Do I re-approve or do you assume the original approval is valid?

**Comment by frankmcsherry** [comment]
I think we do the thing that you previously indicated: once it is accepted I'll take care of it. Not blocked on you! :D

**Comment by frankmcsherry** [comment]
Thanks very much!

**Pull request merged**